PORTF-901 eh8010/20 u-boot unification

yba-siklu opened the pull request:
# Unified EH8020/10 U-Boot
This PR modifies the legacy EH8010 U-Boot (version 20.0.0) to add an SPL that detects the board version and configures the MMDC DDR3 controller according to the DDR3 chip on the board. It detects the NAND size and configures the mtdparts kernel command line accordingly.

Reference: PORTF-901

yba-siklu committed:
PORTF-901 First SPL build - not tested

yba-siklu committed:
Working SPL SPI NOR load

Config 1GB DDR3 for EH8020

Working DMA NAND

Added ccgr_init() to resolve nand_init() hang.

yba-siklu committed:
Fixed board ID detection for PLL init

yba-siklu committed:
Handle mtdparts kernel param per board model

Set the mtdparts kernel command line parameter and U-Boot environment
variable according to board type.

yba-siklu committed:
Incremented UBOOT_MAJOR_STR from 20 to 21

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk_siklu_api.c:
```diff
@@ -6,7 +6,7 @@
 
 
-
+#ifndef CONFIG_SPL_BUILD
```
I think that files that do not build for SPL should be excluded at the `Makefile` level.

yba-siklu replied:
Yes, I thought about this. There are some platforms that do this. I'll write this as a Jira issue.

baruchsiach commented on include/configs/mx6ullevk_siklu.h:
```diff
@@ -18,6 +18,11 @@
 #include "mx6_common.h"
 #include <asm/mach-imx/gpio.h>
 
+#define is_mx6ul_9x9_evk()	CONFIG_IS_ENABLED(TARGET_MX6UL_9X9_EVK)
```
Is that used anywhere?

yba-siklu replied:
Removed.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -28,25 +27,43 @@
+#define SKL_BOARD_TYPE_PCB195 0
+#define SKL_BOARD_TYPE_PCB213 1
+#define SKL_BOARD_TYPE_PCB217 2
+#define SKL_BOARD_TYPE_PCB277 3
```
These macros duplicate the `SKL_BOARD_TYPE_E` enum from `board/freescale/mx6ullevk/siklu_api.h`, and removed in a subsequent commit.

yba-siklu replied:
So what needs to be changed now?

baruchsiach replied:
Modify the commit the needlessly introduced these macros to clean up the history, and tidy affected commits.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -276,12 +285,291 @@ int board_late_init(void)
+/* Called from ./common/board_f.c */
+int board_early_init_f(void)	/* NOT SPL */
+{
+	return 0;
```
Just disable `CONFIG_BOARD_EARLY_INIT_F` to get rid of this empty stub.

yba-siklu replied:
Done.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -276,12 +285,291 @@ int board_late_init(void)
+			mx6_ddr_ioregs	= &eh8010_ddr_ioregs;
+			mx6_grp_ioregs	= &eh8010_grp_ioregs;
+			ddr_sysinfo	= &eh8010_ddr_sysinfo;
+			mx6_mmcd_calib	= &eh8010_mmcd_calib;
```
It looks like `mem_ddr`, `mx6_ddr_ioregs`, and `mx6_grp_ioregs` are the same for both boards. So I think they should share the same structs, with their name indicating that they are shared, maybe something like `eh80x0_mem_ddr`.

yba-siklu replied:
Not clear. What does "are the same for both boards" mean?

baruchsiach replied:
> Not clear. What does "are the same for both boards" mean?

The fields in `eh8010_ddr_ioregs` are initialized exactly the same as in `eh8020_ddr_ioregs`. We can have a single `eh80x0_ddr_ioregs` with these values, and use it for `mx6ul_dram_iocfg()` directly.

yba-siklu replied:
It is likely that I will change these initializations when I do the fine tuning. They are likely to diverge.

baruchsiach replied:
Splitting the struct when there is a need is easy. Removing duplication is hard.

yba-siklu replied:
I tried to remove this duplication and ran into runtime problems that I do not have time to debug. IMHO removing this duplication is a temporary commit, sing the code is likely to diverge again later.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -276,12 +285,291 @@ int board_late_init(void)
+
+int board_spi_cs_gpio(unsigned bus, unsigned cs)
+{
+        return IMX_GPIO_NR(1, 20);
```
This duplicates `board_spi_cs_gpio()` from non-SPL `board/freescale/mx6ullevk/mx6ullevk_siklu_hw.c`. Both routines are now contained within large `#ifdef` code block. Any chance to share code and reduce confusion?

yba-siklu replied:
Fixed

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk_siklu_hw.c:
```diff
@@ -107,32 +111,39 @@ SKL_BOARD_TYPE_E siklu_get_board_type(void)
+		val = ((reg_val & 1<<18)?(1):(0)) + ((reg_val & 1<<19)?(2):(0))
+			+ ((reg_val & 1<<20)?(4):(0))
+			+ ((reg_val & 1<<21)?(8):(0));
+
```
Now I see that this actually duplicates `get_siklu_board_id()` (renamed to `spl_get_siklu_board_id()` in this commit), and all `GPIOx_IOxx` macros. Can we have just one shared routine to detect the board?

yba-siklu replied:
This copy of the routine does things like change the U-Boot environment that we don't do in the SPL. I don't think that there is much added value in extracting common parts of the code. See the comment at the top of this routine regarding the NFS boot.

baruchsiach replied:
It would be nice to have a single routine that just detects the board type, and encapsulates the information of GPIOs listing and order. All other routines can use the shared code without bothering with the GPIO level.

yba-siklu replied:
I tried to do this and ran into runtime problems that I do not have time to debug. It is not work the effort at this time.

baruchsiach commented on cmd/siklu/siklu_nfs_boot.c:
```diff
@@ -27,11 +27,10 @@ void setup_bootargs(const char *bootargs) {
-		mtdparts = CONFIG_SIKLU_DEFAULT_MTD_PARTS;
-		SK_LOG_NFS("Using default mtdparts: %s\n", mtdparts);
+		const char *mtdids;
+		board_mtdparts_default(&mtdids, &mtdparts);
```
Where is `mtdids` used?

baruchsiach replied:
Now I see that `board_mtdparts_default()` is called from generic code.

yba-siklu replied:
Added comments in this file and in board/freescale/mx6ullevk/mx6ullevk.c to clarify.

yba-siklu replied:
Added comments in this file and in board/freescale/mx6ullevk/mx6ullevk.c to clarify.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -27,38 +27,41 @@
+	if (0 == canary) {
+		SKL_BOARD_TYPE_E board_type = siklu_get_board_type();
+
+		snprintf(ids, sizeof(ids), "%s", MTDIDS_DEFAULT);
```
Why copy the string instead of just return a pointer to the `MTDIDS_DEFAULT` string. But this is moot since `mtdids` appears to be unused anyway.

yba-siklu replied:
Fixed - changed to char *

baruchsiach commented on include/configs/mx6ullevk_siklu.h:
```diff
@@ -40,7 +40,21 @@
+ * MTDPARTS_DEFAULT is conditionally checked in cmd/mtdparts.c.
+ * For the unified U-Boot we do not define MTDPARTS_DEFAULT because there is no
+ * single default, it depends on board type. Instead, we set
+ * CONFIG_SYS_MTDPARTS_RUNTIME in configs/mx6ull_14x14_skl_defconfig and we 
```
`CONFIG_SYS_MTDPARTS_RUNTIME` is actually defined a few lines above here.

yba-siklu replied:
Fixed comment.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -27,38 +27,41 @@
-	return id;
+	static canary = 0;
+	static char *siklu_mtdparts;
+	static char ids[24];
```
What is the reason for static variables here? Is it to reduce `siklu_get_board_type()` calls?

yba-siklu replied:
Copied the code structure from siklu_get_board_type() itself to avoid unnecessary assignments and calls to siklu_get_board_type(). Probably doesn't save much.

yba-siklu replied:
Yes, it is to reduce the calls to siklu_get_board_type() to only one call per uptime.

yba-siklu committed:
Remove unnecessary CONFIG_BOARD_EARLY_INIT_F=y

Remove unnecessary CONFIG_BOARD_EARLY_INIT_F=y as per Baruch's review
comment.

yba-siklu committed:
Comment to clarify non-use of mtdids parameter

Added a comment to clarify that the mtdids parameter of
board_mtdparts_default() is not used in setup_bootargs() but is used
in other code that calls board_mtdparts_default().

yba-siklu committed:
Changed char array to char pointer

Changed char array to char pointer as per Baruch's code review comment.

yba-siklu committed:
Corrected comment for CONFIG_SYS_MTDPARTS_RUNTIME

Corrected comment for CONFIG_SYS_MTDPARTS_RUNTIME as per Baruch's code
review.

yba-siklu committed:
Eliminated duplicate board_spi_cs_gpio()

yba-siklu committed:
Removed unused is_mx6ul_9x9_evk macro

Removed unused is_mx6ul_9x9_evk macro as per Baruch's review comment.

yba-siklu committed:
Added comments to explain why code is duplicated

Added comments to explain why the board ID detection routines are
duplicated.

baruchsiach commented on board/freescale/mx6ullevk/mx6ullevk.c:
```diff
@@ -581,12 +581,6 @@ static void ccgr_init(void)
-int board_spi_cs_gpio(unsigned bus, unsigned cs)
-{
-        return IMX_GPIO_NR(1, 20);
-}
```
But now the SPL build gets the empty `board_spi_cs_gpio()` stub in `drivers/spi/mxc_spi.c`. Does that work?

I would be nicer to have a single routine for both SPL and non-SPL.

yba-siklu replied:
"nicer to have". I tried but was not able to get a unified routine working.

baruchsiach commented on include/configs/mx6ullevk_siklu.h:
```diff
@@ -18,15 +18,12 @@
 #include "mx6_common.h"
 #include <asm/mach-imx/gpio.h>
 
-#define is_mx6ul_9x9_evk()	CONFIG_IS_ENABLED(TARGET_MX6UL_9X9_EVK)
```
This and other fixes to commits in the same series should be squashed into the commits they are fixing.

dima1308 commented:
@yba-siklu what about this PR?
We need to complete and merge it before using unified u-boot in the 8010/8020 master of the Porfirio repo.
